// werift/lib/rtp/src/rtp/headerExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeAbsSendTime = exports.serializeTransportWideCC = exports.serializeRepairedRtpStreamId = exports.serializeSdesRTPStreamID = exports.serializeSdesMid = exports.rtpHeaderExtensionsParser = exports.RTP_EXTENSION_URI = void 0;
const src_1 = require("../../../common/src");
exports.RTP_EXTENSION_URI = {
    sdesMid: "urn:ietf:params:rtp-hdrext:sdes:mid",
    sdesRTPStreamID: "urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id",
    repairedRtpStreamId: "urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id",
    transportWideCC: "http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01",
    absSendTime: "http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time",
    dependencyDescriptor: "https://aomediacodec.github.io/av1-rtp-spec/#dependency-descriptor-rtp-header-extension",
    audioLevelIndication: "urn:ietf:params:rtp-hdrext:ssrc-audio-level",
};
function rtpHeaderExtensionsParser(extensions, extIdUriMap) {
    return extensions
        .map((extension) => {
        const uri = extIdUriMap[extension.id];
        switch (uri) {
            case exports.RTP_EXTENSION_URI.sdesMid:
            case exports.RTP_EXTENSION_URI.sdesRTPStreamID:
            case exports.RTP_EXTENSION_URI.repairedRtpStreamId:
                return { uri, value: extension.payload.toString() };
            case exports.RTP_EXTENSION_URI.transportWideCC:
                return { uri, value: extension.payload.readUInt16BE() };
            case exports.RTP_EXTENSION_URI.absSendTime:
                return {
                    uri,
                    value: (0, src_1.bufferReader)(extension.payload, [3])[0],
                };
            case exports.RTP_EXTENSION_URI.audioLevelIndication: {
                const stream = new src_1.BitStream(extension.payload);
                const value = {
                    v: stream.readBits(1) === 1,
                    level: stream.readBits(7),
                };
                return { uri, value };
            }
            default:
                return { uri, value: 0 };
        }
    })
        .reduce((acc, cur) => {
        if (cur)
            acc[cur.uri] = cur.value;
        return acc;
    }, {});
}
exports.rtpHeaderExtensionsParser = rtpHeaderExtensionsParser;
function serializeSdesMid(id) {
    return Buffer.from(id);
}
exports.serializeSdesMid = serializeSdesMid;
function serializeSdesRTPStreamID(id) {
    return Buffer.from(id);
}
exports.serializeSdesRTPStreamID = serializeSdesRTPStreamID;
function serializeRepairedRtpStreamId(id) {
    return Buffer.from(id);
}
exports.serializeRepairedRtpStreamId = serializeRepairedRtpStreamId;
function serializeTransportWideCC(transportSequenceNumber) {
    return (0, src_1.bufferWriter)([2], [transportSequenceNumber]);
}
exports.serializeTransportWideCC = serializeTransportWideCC;
function serializeAbsSendTime(ntpTime) {
    const buf = Buffer.alloc(3);
    const time = (ntpTime >> 14n) & 0x00ffffffn;
    buf.writeUIntBE(Number(time), 0, 3);
    return buf;
}
exports.serializeAbsSendTime = serializeAbsSendTime;
