// werift/lib/webrtc/src/nonstandard/recorder/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaRecorder = void 0;
const rx_mini_1 = __importDefault(require("rx.mini"));
const webm_1 = require("./writer/webm");
class MediaRecorder {
    constructor(path, numOfTracks = 1, options = {}) {
        Object.defineProperty(this, "path", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: path
        });
        Object.defineProperty(this, "numOfTracks", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: numOfTracks
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: options
        });
        Object.defineProperty(this, "writer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "ext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tracks", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "started", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
        Object.defineProperty(this, "onError", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rx_mini_1.default()
        });
        this.ext = path.split(".").slice(-1)[0];
        this.writer = (() => {
            switch (this.ext) {
                case "webm":
                    return new webm_1.WebmFactory(path, options);
                default:
                    throw new Error();
            }
        })();
        this.tracks = options.tracks ?? this.tracks;
        if (this.tracks.length === numOfTracks) {
            this.start().catch((error) => {
                this.onError.execute(error);
            });
        }
    }
    async addTrack(track) {
        this.tracks.push(track);
        await this.start();
    }
    async start() {
        if (this.tracks.length === this.numOfTracks && this.started === false) {
            this.started = true;
            await this.writer.start(this.tracks);
        }
    }
    async stop() {
        await this.writer.stop();
    }
}
exports.MediaRecorder = MediaRecorder;
