// werift/lib/webrtc/src/nonstandard/recorder/writer/webm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebmFactory = void 0;
const promises_1 = require("fs/promises");
const rx_mini_1 = require("rx.mini");
const __1 = require("../../..");
const _1 = require(".");
const sourcePath = "packages/webrtc/src/nonstandard/recorder/writer/webm.ts";
class WebmFactory extends _1.MediaWriter {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "rtpSources", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "unSubscribers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rx_mini_1.EventDisposer()
        });
    }
    async start(tracks) {
        await (0, promises_1.unlink)(this.path).catch((e) => e);
        const inputTracks = tracks.map((track, i) => {
            const trackNumber = i + 1;
            const payloadType = track.codec.payloadType;
            if (track.kind === "video") {
                const codec = (() => {
                    switch (track.codec?.name.toLowerCase()) {
                        case "vp8":
                            return "VP8";
                        case "vp9":
                            return "VP9";
                        case "h264":
                            return "MPEG4/ISO/AVC";
                        case "av1x":
                            return "AV1";
                        default:
                            throw new __1.WeriftError({
                                message: "unsupported codec",
                                payload: { track, path: sourcePath },
                            });
                    }
                })();
                return {
                    kind: "video",
                    codec,
                    clockRate: 90000,
                    trackNumber,
                    width: this.options.width,
                    height: this.options.height,
                    payloadType,
                    track,
                };
            }
            else {
                return {
                    kind: "audio",
                    codec: "OPUS",
                    clockRate: 48000,
                    trackNumber,
                    payloadType,
                    track,
                };
            }
        });
        const webm = new __1.WebmCallback(inputTracks, {
            duration: this.options.defaultDuration ?? 1000 * 60 * 60 * 24,
        });
        const lipsync = new __1.LipsyncCallback();
        this.rtpSources = inputTracks.map(({ track, clockRate, codec }) => {
            const rtpSource = new __1.RtpSourceCallback();
            const rtcpSource = new __1.RtcpSourceCallback();
            track.onReceiveRtp
                .subscribe((rtp) => {
                rtpSource.input(rtp.clone());
            })
                .disposer(this.unSubscribers);
            track.onReceiveRtcp
                .subscribe((rtcp) => {
                rtcpSource.input(rtcp);
            })
                .disposer(this.unSubscribers);
            const ntpTime = new __1.NtpTimeCallback(clockRate);
            if (track.kind === "video") {
                const depacketizer = new __1.DepacketizeCallback(codec, {
                    isFinalPacketInSequence: (h) => h.marker,
                });
                const jitterBuffer = new __1.JitterBufferCallback(clockRate);
                rtpSource.pipe(jitterBuffer.input);
                rtcpSource.pipe(ntpTime.input);
                jitterBuffer.pipe(ntpTime.input);
                ntpTime.pipe(depacketizer.input);
                depacketizer.pipe(lipsync.inputVideo);
                lipsync.pipeVideo(webm.inputVideo);
            }
            else {
                const depacketizer = new __1.DepacketizeCallback(codec);
                rtpSource.pipe(ntpTime.input);
                rtcpSource.pipe(ntpTime.input);
                ntpTime.pipe(depacketizer.input);
                depacketizer.pipe(lipsync.inputAudio);
                lipsync.pipeAudio(webm.inputAudio);
            }
            return rtpSource;
        });
        webm.pipe((0, __1.saveToFileSystem)(this.path));
    }
    async stop() {
        await Promise.all(this.rtpSources.map((r) => r.stop()));
        this.unSubscribers.dispose();
    }
}
exports.WebmFactory = WebmFactory;
const supportedVideoCodecs = ["h264", "vp8", "vp9", "av1x"];
