// werift/lib/rtp/src/processor/ntpTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NtpTimeBase = void 0;
const crypto_1 = require("crypto");
const __1 = require("..");
class NtpTimeBase {
    constructor(clockRate) {
        Object.defineProperty(this, "clockRate", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: clockRate
        });
        Object.defineProperty(this, "baseNtpTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "baseRtpTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "latestNtpTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "latestRtpTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "currentElapsed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "buffer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "internalStats", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {}
        });
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (0, crypto_1.randomUUID)()
        });
        Object.defineProperty(this, "payloadType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
    }
    toJSON() {
        return {
            baseRtpTimestamp: this.baseRtpTimestamp,
            latestRtpTimestamp: this.latestRtpTimestamp,
            baseNtpTimestamp: this.baseNtpTimestamp && (0, __1.ntpTime2Sec)(this.baseNtpTimestamp),
            latestNtpTimestamp: this.latestNtpTimestamp && (0, __1.ntpTime2Sec)(this.latestNtpTimestamp),
            bufferLength: this.buffer.length,
            currentElapsed: this.currentElapsed,
            id: this.id,
            clockRate: this.clockRate,
            ...this.internalStats,
        };
    }
    stop() {
        this.buffer = [];
        this.internalStats = {};
    }
    processInput({ rtcp, rtp, eol }) {
        if (eol) {
            this.stop();
            return [{ eol: true }];
        }
        if (rtcp && rtcp instanceof __1.RtcpSrPacket) {
            const { ntpTimestamp, rtpTimestamp } = rtcp.senderInfo;
            this.latestNtpTimestamp = ntpTimestamp;
            this.latestRtpTimestamp = rtpTimestamp;
            if (this.baseNtpTimestamp == undefined) {
                this.baseNtpTimestamp = ntpTimestamp;
                this.baseRtpTimestamp = rtpTimestamp;
            }
        }
        if (rtp) {
            this.buffer.push(rtp);
            this.payloadType = rtp.header.payloadType;
            const res = [];
            this.buffer = this.buffer
                .map((rtp) => {
                const ntp = this.updateNtp(rtp.header.timestamp);
                if (ntp != undefined) {
                    const ms = ntp * 1000;
                    res.push({ rtp, time: Math.round(ms) });
                    return undefined;
                }
                return rtp;
            })
                .filter((r) => r != undefined);
            return res;
        }
        return [];
    }
    /**
     *
     * @param rtpTimestamp
     * @returns sec
     */
    calcNtp({ rtpTimestamp, baseNtpTimestamp, baseRtpTimestamp, elapsedOffset, }) {
        const rotate = Math.abs(rtpTimestamp - baseRtpTimestamp) > (__1.Max32Uint / 4) * 3;
        const elapsed = rotate
            ? rtpTimestamp + __1.Max32Uint - baseRtpTimestamp
            : rtpTimestamp - baseRtpTimestamp;
        const elapsedSec = elapsed / this.clockRate;
        this.internalStats["totalElapsed"] =
            (this.internalStats["totalElapsed"] ?? 0) + elapsedSec;
        // sec
        const ntp = (0, __1.ntpTime2Sec)(baseNtpTimestamp) + elapsedOffset + elapsedSec;
        return { ntp, elapsedSec };
    }
    updateNtp(rtpTimestamp) {
        if (this.baseRtpTimestamp == undefined ||
            this.baseNtpTimestamp == undefined ||
            this.latestNtpTimestamp == undefined ||
            this.latestRtpTimestamp == undefined) {
            return;
        }
        this.internalStats["inputRtp"] = rtpTimestamp;
        const base = this.calcNtp({
            rtpTimestamp,
            baseNtpTimestamp: this.baseNtpTimestamp,
            baseRtpTimestamp: this.baseRtpTimestamp,
            elapsedOffset: this.currentElapsed,
        });
        const latest = this.calcNtp({
            rtpTimestamp,
            baseNtpTimestamp: this.latestNtpTimestamp,
            baseRtpTimestamp: this.latestRtpTimestamp,
            elapsedOffset: 0,
        });
        this.internalStats["calcBaseNtp"] = base.ntp;
        this.internalStats["calcLatestNtp"] = latest.ntp;
        if (base.ntp < latest.ntp) {
            // update baseNtp
            this.baseNtpTimestamp = this.latestNtpTimestamp;
            this.baseRtpTimestamp = this.latestRtpTimestamp;
            this.currentElapsed = 0;
            this.internalStats["calcNtp"] = latest.ntp;
            return latest.ntp;
        }
        else {
            this.currentElapsed += base.elapsedSec;
            this.baseRtpTimestamp = rtpTimestamp;
            this.internalStats["calcNtp"] = base.ntp;
            return base.ntp;
        }
    }
}
exports.NtpTimeBase = NtpTimeBase;
